### PR TITLE
Preserve Coordinates when Indexing UxDataArray

### LINKED ISSUE #945: `_slice_from_grid` may need to use coords from a sliced dataarray
### Version

v2024.08.2

### How did you install UXarray?

Conda

### What happened?

`_slice_from_grid` in UxDataArray create a new subset dataarray using coordinates info from the original dataarray as shown in the lines below:
https://github.com/UXARRAY/uxarray/blob/f71e11bef0fabe3ac74767767ad0509cef5cf504/uxarray/core/dataarray.py#L1099-L1106

This does not cause an issue when the original dataarray does not have coordinates that are sliced, but the constructor fails otherwise with a conflicting size for dimension. Probably a better or a right approach is creating a new sliced dataarray with coordinates from the sliced `d_var` not from the original `self`.

### What did you expect to happen?

A new sliced (subset) dataarray needs to be created with a correctly sliced coordinate variables, not from the original unsliced coordinate variables.

### Can you provide a MCVE to repoduce the bug?

_No response_

### PR DESCRIPTION
**Pull request by philipc2**
Closes #945

## Overview

* Preserves coordinates when indexing a `UxDataArray`

### CODE REVIEW

**Review by ahijevyc — COMMENTED** [review]
Looks good. I tested the change and tried to replicate the issue I had #818 with the uxarray.UxDataArray.Indexes disappearing but had no problems. 

However, given this fix, I can think of two other places a similar change might be needed:
 https://github.com/UXARRAY/uxarray/blob/ec5cc8f279ae7026e5f42ab0570105cd6861ced7/uxarray/remap/nearest_neighbor.py#L110

and 

https://github.com/UXARRAY/uxarray/blob/ec5cc8f279ae7026e5f42ab0570105cd6861ced7/uxarray/remap/inverse_distance_weighted.py#L151

**Comment by philipc2** [comment]
> Looks good. I tested the change and tried to replicate the issue I had #818 with the uxarray.UxDataArray.Indexes disappearing but had no problems.
> 
> However, given this fix, I can think of two other places a similar change might be needed:
> 
> https://github.com/UXARRAY/uxarray/blob/ec5cc8f279ae7026e5f42ab0570105cd6861ced7/uxarray/remap/nearest_neighbor.py#L110
> 
> and
> 
> https://github.com/UXARRAY/uxarray/blob/ec5cc8f279ae7026e5f42ab0570105cd6861ced7/uxarray/remap/inverse_distance_weighted.py#L151

Good catches, I'll update these also.

**Comment by philipc2** [comment]
@ahijevyc 

When remapping, should the coordinates from the original data array be dropped? It wouldn't make much sense to preserve them (unless I'm understanding it wrong)

**Review by ahijevyc — COMMENTED** [review]
The _spatial_ coordinates should not be transferred, but if there are _non-spatial_ coordinates like `time`, `ensemble_member`, or `forecast_hour`, then those should be transferred to the destination. I think that's why I originally thought the `coords` argument was necessary.

**Comment by philipc2** [comment]
> The _spatial_ coordinates should not be transferred, but if there are _non-spatial_ coordinates like `time`, `ensemble_member`, or `forecast_hour`, then those should be transferred to the destination. I think that's why I originally thought the `coords` argument was necessary.

That makes sense! I wasn't considering the non-spatial cases.

File: uxarray/remap/inverse_distance_weighted.py [hunk 1]
```diff
@@ -148,7 +148,6 @@ def _inverse_distance_weighted_remap_uxda(
     uxda_remap = uxarray.core.dataarray.UxDataArray(
         data=destination_data,
         name=source_uxda.name,
-        coords=source_uxda.coords,
```
Comment (ahijevyc) [inline uxarray/remap/inverse_distance_weighted.py:151]: @philipc2 Here's my solution to preserve any non-spatial coordinates (like time or ensemble_member) while dropping the spatial one:

Copy `source_uxda.coords` to a new `xarray.Coordinates` object. Then delete the key associated with the spatial dimension. Then use that new object as the coords argument to `uxarray.core.dataarray.UxDataArray`.

```
    destination_coords = source_uxda.coords
    del(destination_coords[destination_dim])

    # construct data array for remapping variable
    uxda_remap = uxarray.core.dataarray.UxDataArray(
        data=destination_data,
        name=source_uxda.name,
        coords=destination_coords,
```
Does this make sense?

**Review by ahijevyc — CHANGES_REQUESTED** [review]
I provided a way to drop the remapped coordinate (which I agree, should not get transferred from the source) and still preserve any non-spatial coordinates.

File: uxarray/remap/inverse_distance_weighted.py (hunk 1 above)
Comment (ahijevyc) [reply uxarray/remap/inverse_distance_weighted.py:151]: @philipc2 Thanks .  Just to make sure I understand, does the _preserve_coordinates function slice along the spatial coordinate and return the sliced spatial coordinate (along with the non-spatial coordinates)? If so, that's great. I don't think my solution did that. My solution simply deleted the spatial coordinate. Retaining the properly-sliced spatial coordinate is good.  
Therefore, I don't think we really need a `dims_to_drop` option.  When I dropped it in my suggested code, I was just being lazy. I couldn't think of a way to provide the sliced spatial coordinates. We don't want to completely drop the spatial dimension, just slice it.  So could the `dims_to_drop` option be removed?

**Review by ahijevyc — COMMENTED** [review]
I had a question about the new util function and whether it can be simplified.

File: uxarray/remap/inverse_distance_weighted.py (hunk 1 above)
Comment (philipc2) [reply uxarray/remap/inverse_distance_weighted.py:151]: I've updated the logic in the `UxDataArray.isel()` function, which should now correctly handle slicing the spatial dimensions.

As for the remapping, I've re-added the deletion approach you mentioned, with a few extra checks.

**Review by ahijevyc — CHANGES_REQUESTED** [review]
Looks good. 
I couldn't think of a reason to find and delete the grid dimension found in the list ["n_node", "n_edge", "n_face"] when one could use simply delete `destination_dim`, which was defined earlier. So I removed a couple for-block and if-block levels.

**Review by ahijevyc — APPROVED** [review]
I approve

**Pull request merged**